[System.Runtime.Loader] Add hot reload test infrastructure

lambdageek opened the pull request:
Adding infrastructure for hot reload testing.

For each test we define a new library assembly project.  The `.csproj` has a
`DeltaScript` property that specifies a JSON file that lists the name of an
initial source file, and a list of updated versions of that file.  The
`hotreload-delta-gen` tool runs during the build to read the delta script and
create deltas that incorporate the updates.

The main testsuite references all the test assemblies, and when a test runs, it
calls `ApplyUpdateUtil.ApplyUpdate` to load subsequent deltas
and then compares the results before and after an update.

Dependencies:

- https://github.com/dotnet/hotreload-utils  - there is now a `Microsoft.DotNet.HotReload.Utils.Generator.BuildTool` MSBuild targets nupkg on the `dotnet6-transport` nuget feed.
- If `DOTNET_MODIFIABLE_ASSEMBLIES` is not set in the environment, we use remote executor to run the tests.

Mono is technically enabled, but practically it's not running anywhere:
- We have not enabled the MonoMetadataUpdate support on desktop linux or macos
- For wasm, ios and Android since we don't have remote executor support, we would need to adjust the test pipelines to pass the `DOTNET_MODIFIABLE_ASSEMBLIES` flag via xharness (and the app builder tasks).
- For ios and Android we would need to enable the interpreter as the execution engine, otherwise the tests are skipped.


To try it out locally with CoreCLR run:

```
./dotnet.sh build src/libraries/System.Runtime.Loader/tests /t:Test
```

For mono, build the runtime with `/p:MonoMetadataUpdate=true` and setenv `MONO_ENV_OPTIONS=--interp`

lambdageek committed:
[System.Runtime.Loader] Add hot reload test infrastructure

Adding infrastructure for hot reload testing.

For each test we define a new library assembly project.  The `.csproj` has a
`DeltaScript` property that specifies a JSON file that lists the name of an
initial source file, and a list of updated versions of that file.  The
`hotreload-delta-gen` tool runs during the build to read the delta script and
create deltas that incorporate the updates.

The main testsuite references all the test assemblies, and when a test runs, it
calls `ApplyUpdateUtil.ApplyUpdate` to load subsequent deltas
and then compares the results before and after an update.

Dependencies:

- https://github.com/dotnet/hotreload-utils  the `hotreload-delta-gen` binary
must be installed and on the PATH  (TODO: package it as a dotnet tool and
publish to a transport nuget package)

Needs work:

- Mono test runs need to pass the `FEATURE_MONO_METADATA_UPDATE` property to
  msbuild to the test project.  This is because not every mono configuration
  includes hot reload support.  Eventually this should be subsumed by the
  runtime API to querty hot reload capabilities https://github.com/dotnet/runtime/issues/50111
- All runs need to pass `DOTNET_MODIFIABLE_ASSEMBLIES=debug` to the testsuite.
  Hot reload only works when:
  - the assemblies are compiled with the Debug compiler setting,
  - and, if the runtime is tarted with the above environment variable set.
- The GenerateHotReloadDeltas.targets needs some output file work to run `hotreload-delta-gen` less
  frequently.  Right now it runs every time even if everything is up to date.
  For CI testing we should ideally compile the deltas ahead of time once.

To try it out locally:

1. checkout and build `hotreload-utils` and do `dotnet publish --self-contained -r <RID>` to
put `hotreload-delta-gen` into that projects' `artifacts/...` forlder.  Add the
publish folder to your `$PATH`

2. In dotnet/runtime run
```
DOTNET_MODIFIABLE_ASSEMBLIES=debug MONO_ENV_OPTIONS=--interp ./dotnet.sh build src/libraries/System.Runtime.Loader/tests /p:MonoMetadataUpdate=true /t:Test
```

(CoreCLR doesn't need `MONO_ENV_OPTIONS` or `MonoMetadataUpdate`)

lambdageek commented:
@tommcdon @mikem8361 So here's the general shape of how I think hot reload tests should look.

@akoeplinger @ViktorHofer I really need help setting up arcade stuff for https://github.com/dotnet/hotreload-utils - the end state is that I'd like for `hotreload-delta-gen` to be a dotnet tool that's packaged in a transport package that dotnet/runtime could pick up using arcade/BAR/maestro.  I tried setting up the eng/ folder, and Alexander set up an internal repo for the publishing.  But there's no pipeline that pushes out packages yet.

@akoeplinger @ViktorHofer I also will need some help with CI.  The tests depend on some environment variable being set (`DOTNET_MODIFIABLE_ASSEMBLIES=debug`) and I'm not sure if we should set that in the pipeline, or if there's some way to have the test project itself start up with environment values set.

mikem8361 commented:
One thing to note that "source build" will require the hotreload-utils repo to built as part of it.  I'm not sure exactly how tools like this are handled as part of the source build.

jkoritzinsky commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.targets:
```diff
@@ -0,0 +1,81 @@
+
+  <!-- Computes the names of the files that hotreload-delta-gen will produce, given the name of the base assembly and the
+       number of deltas -->
+  <UsingTask TaskName="ComputeDeltaFileOutputNames" TaskFactory="RoslynCodeTaskFactory"
```
Can we actually create a tasks assembly to use here (probably in `src/tasks`) instead of using the RoslynCodeTaskFactory?

lambdageek replied:
Added a `GenerateHotReloadDelta.BuildTasks` assembly next to the tests.  Also took the opportunity to compute the number of expected deltas directly from the json script, instead of requiring a (redundant) DeltaCount property in the projects. Should  improve usability a bit.

lambdageek commented:
> One thing to note that "source build" will require the hotreload-utils repo to built as part of it. I'm not sure exactly how tools like this are handled as part of the source build.

Good point.  Not sure. I think the situation will be similar to how our LLVM fork, or the other transport nugets are built.  (Possibly with a bit more of a bootstrapping problem since hotreload-utils depends on Roslyn which depends on the runtime)

lambdageek committed:
Make a proper task for computing hotreload-delta-gen output files

lambdageek committed:
Don't need DeltaCount property, compute from json

Compute the number of deltas that hotreload-delta-gen will produce by
parsing the json script and counting the number of changes.

lambdageek committed:
Add dependency on hotreload-delta-gen tool

lambdageek committed:
use 'dotnet tool run hotreload-delta-gen' to generate EnC deltas

lambdageek committed:
Use remote executor if DOTNET_MODIFIABLE_ASSEMBLIES is not set

lambdageek committed:
Use DotNetTool property to run hotreload-delta-gen

just `dotnet tool run hotreload-delta-gen` can fail on 6.0 preview 2 because of
borked argument parsing

lambdageek committed:
Don't run on Mono for now

1. mobile devices and wasm will need xharness or the app builder tasks to set the
DOTNET_MODIFIABLE_ASSEMBLIES environment variable.
2. for console apps, we need the hot reload capabilities API to check if the
current runtime can do updates

lambdageek commented:
various Libraries builds are failing with 

> Run "dotnet tool restore" to make the "hotreload-delta-gen" command available.

@ViktorHofer @akoeplinger is it not enough to add the reference to eng/Version.Details.xml and to .config/dotnet-tools.json?  Is there something else I need to tweak?

lambdageek committed:
bump hotreload-delta-gen package version

lambdageek commented:
@ViktorHofer ah ok... I see this is due to 

https://github.com/dotnet/runtime/blob/9f5568993103f88b48df4aa0b6f639d64c4d0af9/eng/Tools.props#L3-L5

and ultimately due to https://github.com/dotnet/sdk/issues/10938 - which was closed - does that mean its ok to restore the tools on CI? 

So... what do I do?   I don't need the hotreload-delta-gen tool on helix, but I need it to run when we're creating the archives for the tests (or, of course for non-CI scenarios when we're just running a testsuite locally).

Is there some way to identify that we're on CI but just doing the package build, not the helix fanout of the test runs?

lambdageek committed:
Install the CLI tools in CI

This was previously disabled due to https://github.com/dotnet/sdk/issues/10938
and its continuation issue https://github.com/dotnet/arcade/issues/5402

But things might be better now?

lambdageek committed:
Run tests on Mono if feature enabled and interp is used

lambdageek committed:
fix remote executor detection

just calling RemoteExecutor.IsSupported on wasm throws

```
System.TypeInitializationException: The type initializer for 'Microsoft.DotNet.RemoteExecutor.RemoteExecutor' threw an exception.
---> System.PlatformNotSupportedException: System.Diagnostics.Process
is not supported on this platform.
```

lambdageek committed:
remove unneeded DefineConstants

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.targets:
```diff
@@ -0,0 +1,45 @@
+  <!-- Have to be careful about AfterTargets. If we happen to go after a design-time target, we could get into an infinite loop. -->
+  <Target Name="CompileDiff" AfterTargets="Build" Condition="'$(DeltaScript)' != ''">
+    <PropertyGroup>
+      <HotreloadDeltaGenCommand>"$(DotNetTool)" hotreload-delta-gen</HotreloadDeltaGenCommand>
```
Presumably this adds a source build dependency onto hotreload-delta-gen. Is that repo sourcebuild compatible? Today we don't use any dotnet (repo) tools as part of the product build and I would like to better understand why this can't be a normal nuget package instead? Additional thoughts on this:

1. This adds an out-of-proc invocation which slows down the build
2. Using a dotnet tool requires them (the manifest) to be restored upfront. This works against the goal to make any project in the repository invokable without any up-front steps. If this would be a nuget package with a targets file in it instead, we would have a strong dependency and the package would only be restored if this project is restored.
3. When building other parts of the repository, this tool always needs to be restored.

If we control the tool ourselves then I would prefer to consume a nuget package with an msbuild targets file in it which then invokes a task to avoid the out-of-proc hop. Is that possible?

cc @ericstj

lambdageek replied:
> an msbuild targets file in it which then invokes a task to avoid the out-of-proc hop. Is that possible

This won't work - hotreload-delta-gen needs to open the current project as a roslyn MSBuildWorkspace - which throws an exception because the current project already has MSBuild loaded.

ViktorHofer replied:
If out-of-process isn't negotiable then I'm ok with that. Please consider my other suggestions as I believe a dependency on a dotnet tool isn't great.

lambdageek replied:
> If out-of-process isn't negotiable then I'm ok with that. Please consider my other suggestions as I believe a dependency on a dotnet tool isn't great.

@ViktorHofer okay, I'll give it a try.  Is this blog post a good guide to how to package a build tool in a nuget? https://natemcmaster.com/blog/2017/11/11/build-tools-in-nuget/  Is there anything special I need to do differnetly for our repos?

ViktorHofer replied:
Some of what's said in this blog doesn't apply anymore but in general this is a good guidance, yes. https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.ApiCompat/src/Microsoft.DotNet.ApiCompat.csproj is an example of how you would package such a tool in Arcade which handles most of what you need via the BuildTask.targets file which is imported in the project.

In general creating a nuget package which invokes an executable via a targets file underneath isn't too hard. Give it a try and I'm happy to review your changes where-ever they need to be done.

lambdageek replied:
Done.  Thanks for your help @ViktorHofer - I switched this PR over to using the Generator.BuildTool nupkg and removed the tool from the manifest.

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.targets:
```diff
@@ -0,0 +1,45 @@
+
+  <!-- Don't invoke hotreload-delta-gen, but compute the names of the artifacts that it would have produced. -->
+  <Target Name="ComputeDeltaFileOutputNames" Condition="'$(DeltaScript)' != ''">
+    <!-- FIXME: is AssemblyName the best property to get the output assembly? -->
+    <ComputeDeltaFileOutputNames BaseAssemblyName="$(OutputPath)\$(AssemblyName).dll" DeltaScript="$(DeltaScript)">
```
```suggestion
    <ComputeDeltaFileOutputNames BaseAssemblyName="$(TargetPath)" DeltaScript="$(DeltaScript)">
```

lambdageek committed:
Use `$(TargetPath)` as input to ComputeDeltaOutputNames

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.targets:
```diff
@@ -0,0 +1,45 @@
+  <Target Name="ContentForDeltaFileOutputNames"
+    DependsOnTargets="ComputeDeltaFileOutputNames"
+    BeforeTargets="AssignTargetPaths"
+    Condition="'$(DesignTimeBuild)' != 'true' and '$(BuildingProject)' == 'true'"
```
I believe BuildingProject is a legacy property from the old CPS and shouldn't be used anymore: https://github.com/dotnet/msbuild/issues/2417

lambdageek committed:
drop "BuildingProject" property check

It's for the old project style, only

lambdageek committed:
Use Microsoft.DotNet.HotReload.Utils.Generator.BuildTool instead of hotreload-delta-gen

Use an msbuild target nuget instead of the CLI tool

lambdageek committed:
Use well known version for Generator BuiltTool version

lambdageek committed:
Revert "Install the CLI tools in CI"

Don't need this anymore - using an MSBuild targets nuget now instead of a
dotnet tool

This reverts commit c125be86a3324d4bbcce1728b124fb0b4abd9add.

lambdageek committed:
Use published Generator.BuildTool package

danmoseley commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs:
```diff
@@ -0,0 +1,43 @@
+    /// Each test Foo has a corresponding assembly under
+    /// System.Reflection.Metadata.ApplyUpate.Test.Foo The Foo.csproj has a delta
+    /// script that applies one or more updates to Foo.dll The ApplyUpdateTest
+    /// testusuite runs each test in sequence, loading the corresponding
```
tyop

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/ComputeDeltaFileOutputNames.cs:
```diff
@@ -0,0 +1,46 @@
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public class ComputeDeltaFileOutputNames : Microsoft.Build.Utilities.Task {
```
nit formatting:

```suggestion
public class ComputeDeltaFileOutputNames : Microsoft.Build.Utilities.Task
{
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/ComputeDeltaFileOutputNames.cs:
```diff
@@ -0,0 +1,46 @@
+using System;
```
missing copyright header

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/GenerateHotReloadDelta.BuildTasks.csproj:
```diff
@@ -0,0 +1,15 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Could / should this build task also be part of the `Microsoft.DotNet.HotReload.Utils.Generator.BuildTool` package?

lambdageek replied:
Yes. I would prefer to move it in a follow-up PR.

lambdageek replied:
Ok, actually moved the rest of the build logic in the Generator.BuiltTool package.  Also added incremental rebuild support.

@ViktorHofer mind taking one more look at this PR

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/DeltaScriptParser.cs:
```diff
@@ -0,0 +1,41 @@
+        return json;
+    }
+
+    public class Script {
```
```suggestion
    public class Script
    {
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/DeltaScriptParser.cs:
```diff
@@ -0,0 +1,41 @@
+    }
+
+
+    public class Change {
```
```suggestion
    public class Change
    {
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/DeltaScriptParser.cs:
```diff
@@ -0,0 +1,41 @@
+using System;
```
missing copyright header

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/ComputeDeltaFileOutputNames.cs:
```diff
@@ -0,0 +1,46 @@
+            return false;
+        }
+        ITaskItem[] result = new TaskItem[3*count];
+        for (int i = 0; i < count; ++i) {
```
```suggestion
        for (int i = 0; i < count; ++i)
        {
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/ComputeDeltaFileOutputNames.cs:
```diff
@@ -0,0 +1,46 @@
+        try {
+            var json = DeltaScriptParser.Parse (DeltaScript).Result;
+            count = json.Changes.Length;
+        } catch (System.Text.Json.JsonException exn) {
```
```suggestion
        } catch (System.Text.Json.JsonException exn)
        {
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/ComputeDeltaFileOutputNames.cs:
```diff
@@ -0,0 +1,46 @@
+        }
+        string baseAssemblyName = BaseAssemblyName;
+        int count;
+        try {
```
```suggestion
        try
        {
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/GenerateHotReloadDelta.BuildTasks/ComputeDeltaFileOutputNames.cs:
```diff
@@ -0,0 +1,46 @@
+
+    public override bool Execute()
+    {
+        if (!System.IO.File.Exists (DeltaScript)) {
```
```suggestion
        if (!System.IO.File.Exists (DeltaScript))
        {
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdate/System.Reflection.Metadata.ApplyUpdate.Test.MethodBody1/System.Reflection.Metadata.ApplyUpdate.Test.MethodBody1.csproj:
```diff
@@ -0,0 +1,14 @@
+  </ItemGroup>
+
+  <PropertyGroup>
+    <DeltaScript>deltascript.json</DeltaScript>
```
nit: merge this propertygroup with the other one at the top

lambdageek replied:
done

lambdageek committed:
fix typos, indentantion; add copyright headers

lambdageek committed:
Remove ComputeDeltaFileOutputNames use nuget version

The Microsoft.DotNet.HotReload.Utils.Generator.BuildTool nuget now includes a
task to compute the output names.  So the targets in the nuget are entirely
self-contained now - incremental builds and project references should work now.

Bump to version 1.0.1 of Generator.BuildTool which has the necessary targets

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs:
```diff
@@ -0,0 +1,43 @@
+                Assert.Equal ("NEWEST STRING", r);
+            });
+        }
+        
```
nit:
```suggestion
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs:
```diff
@@ -0,0 +1,43 @@
+        [Fact]
+        void StaticMethodBodyUpdate()
+        {
+            ApplyUpdateUtil.TestCase(static () => {
```
```suggestion
            ApplyUpdateUtil.TestCase(static () =>
            {            
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdate/System.Reflection.Metadata.ApplyUpdate.Test.MethodBody1/MethodBody1_v2.cs:
```diff
@@ -0,0 +1,9 @@
+
```
missing license header

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdate/System.Reflection.Metadata.ApplyUpdate.Test.MethodBody1/MethodBody1_v1.cs:
```diff
@@ -0,0 +1,9 @@
+
```
missing license header

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdate/System.Reflection.Metadata.ApplyUpdate.Test.MethodBody1/MethodBody1.cs:
```diff
@@ -0,0 +1,9 @@
+
```
missing license header

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateUtil.cs:
```diff
@@ -0,0 +1,148 @@
+                Console.Error.WriteLine ($"Running test using RemoteExecutor");
+                AddRemoteInvokeOptions(ref options);
+                RemoteExecutor.Invoke(testBody, options).Dispose();
+            } else {
```
```suggestion
            }
            else
            {
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateUtil.cs:
```diff
@@ -0,0 +1,148 @@
+        public static void TestCase(Action testBody,
+                                    RemoteInvokeOptions options = null)
+        {
+            if (UseRemoteExecutor) {
```
```suggestion
            if (UseRemoteExecutor)
            {
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateUtil.cs:
```diff
@@ -0,0 +1,148 @@
+#endif
+        }
+
+        private static System.Collections.Generic.Dictionary<Assembly, int> assembly_count = new ();
```
```suggestion
        private static System.Collections.Generic.Dictionary<Assembly, int> assembly_count = new();
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateUtil.cs:
```diff
@@ -0,0 +1,148 @@
+
+namespace System.Reflection.Metadata
+{
+
```
```suggestion
```

ViktorHofer commented on src/libraries/System.Runtime.Loader/tests/ApplyUpdateTest.cs:
```diff
@@ -0,0 +1,43 @@
+    [ConditionalClass(typeof(ApplyUpdateUtil), nameof (ApplyUpdateUtil.IsSupported))]
+    public class ApplyUpdateTest
+    {
+
```
```suggestion
```

ViktorHofer reviewed (approved):
LGTM modulo style nits

lambdageek committed:
Fix style nits; add more copyright headers
